File: weather_cli/providers/wttr.py
import json
import urllib.error
import urllib.request
from weather_cli.providers.base import BaseWeatherProvider
from weather_cli.providers.models import HourlyPoint, NormalizedWeatherData

WWO_TO_WMO = {
    113: 0,   # Sunny/Clear
    116: 2,   # Partly Cloudy
    119: 3,   # Cloudy
    122: 3,   # Overcast
    143: 45,  # Mist
    176: 61,  # Patchy rain nearby
    179: 71,  # Patchy snow nearby
    182: 71,  # Patchy sleet nearby
    185: 56,  # Patchy freezing drizzle nearby
    200: 95,  # Thundery outbreaks nearby
    227: 77,  # Blowing snow
    230: 75,  # Blizzard
    248: 45,  # Fog
    260: 48,  # Freezing fog
    263: 51,  # Patchy light drizzle
    266: 51,  # Light drizzle
    281: 56,  # Freezing drizzle
    284: 57,  # Heavy freezing drizzle
    293: 61,  # Patchy light rain
    296: 61,  # Light rain
    299: 63,  # Moderate rain at times
    302: 63,  # Moderate rain
    305: 65,  # Heavy rain at times
    308: 65,  # Heavy rain
    311: 66,  # Light freezing rain
    314: 67,  # Moderate or heavy freezing rain
    317: 71,  # Light sleet
    320: 75,  # Moderate or heavy sleet
    323: 71,  # Patchy light snow
    326: 71,  # Light snow
    329: 73,  # Patchy moderate snow
    332: 73,  # Moderate snow
    335: 75,  # Patchy heavy snow
    338: 75,  # Heavy snow
    350: 77,  # Ice pellets
    353: 80,  # Light rain shower
    356: 81,  # Moderate or heavy rain shower
    359: 82,  # Torrential rain shower
    362: 85,  # Light sleet showers
    365: 86,  # Moderate or heavy sleet showers
    368: 85,  # Light snow showers
    371: 86,  # Moderate or heavy snow showers
    374: 77,  # Light showers of ice pellets
    377: 77,  # Moderate or heavy showers of ice pellets
    386: 95,  # Patchy light rain with thunder
    389: 95,  # Moderate or heavy rain with thunder
    392: 96,  # Patchy light snow with thunder
    395: 99,  # Moderate or heavy snow with thunder
}


def _wwo_code_to_wmo(wwo_code: int) -> int:
    return WWO_TO_WMO.get(wwo_code, 0)


class WttrProvider(BaseWeatherProvider):
    @property
    def name(self) -> str:
        return "wttr.in"

    @property
    def provider_name(self) -> str:
        return "wttr.in"

    def fetch(self, lat: float, lon: float, start_date: str, end_date: str) -> NormalizedWeatherData:
        # wttr.in format=j1 URL format
        url = f"https://wttr.in/{lat},{lon}?format=j1"
        
        try:
            with urllib.request.urlopen(url) as resp:
                raw_data = json.loads(resp.read().decode())
        except urllib.error.URLError as e:
            raise ConnectionError(f"Failed to reach wttr.in weather service: {e}") from e

        return self._normalize(raw_data, start_date, end_date)

    def _normalize(self, raw: dict, start_date: str, end_date: str) -> NormalizedWeatherData:
        weather_list = raw.get("weather", [])
        
        points = []
        for day in weather_list:
            day_date = day.get("date", "")
            # Filter by date range (inclusive)
            if not day_date or not (start_date <= day_date <= end_date):
                continue
                
            hourly_blocks = day.get("hourly", [])
            if not hourly_blocks:
                continue
                
            # Expand 3-hourly blocks to 24-hourly points (00:00 to 23:00)
            for h in range(24):
                block_idx = min(h // 3, len(hourly_blocks) - 1)
                block = hourly_blocks[block_idx]
                
                temp = block.get("tempC", "0")
                feels_like = block.get("FeelsLikeC", temp)
                precip_prob = block.get("chanceofrain", "0")
                precip = block.get("precipMM", "0.0")
                humidity = block.get("humidity", "0")
                wind_speed = block.get("windspeedKmph", "0")
                wind_dir = block.get("winddirDegree", "0")
                cloud_cover = block.get("cloudcover", "0")
                wwo_code = int(block.get("weatherCode", "113"))
                
                point = HourlyPoint(
                    time=f"{day_date}T{h:02d}:00",
                    temperature=float(temp),
                    apparent_temperature=float(feels_like),
                    precipitation_probability=float(precip_prob),
                    precipitation=float(precip),
                    humidity=float(humidity),
                    wind_speed=float(wind_speed),
                    wind_direction=float(wind_dir),
                    cloud_cover=float(cloud_cover),
                    weather_code=_wwo_code_to_wmo(wwo_code),
                )
                points.append(point)
                
        return NormalizedWeatherData(
            provider_name=self.provider_name,
            hourly=points
        )

File: src/providers/wttr.rs
use serde::Deserialize;
use crate::providers::base::{BaseWeatherProvider, FetchContext};
use crate::providers::models::{HourlyPoint, NormalizedWeatherData};

#[derive(Clone)]
pub struct WttrProvider;

#[derive(Deserialize, Debug)]
struct WttrHourlyBlock {
    #[serde(rename = "tempC")]
    temp_c: Option<String>,
    #[serde(rename = "FeelsLikeC")]
    feels_like_c: Option<String>,
    chanceofrain: Option<String>,
    #[serde(rename = "precipMM")]
    precip_mm: Option<String>,
    humidity: Option<String>,
    #[serde(rename = "windspeedKmph")]
    wind_speed_kmh: Option<String>,
    #[serde(rename = "winddirDegree")]
    wind_dir_degree: Option<String>,
    cloudcover: Option<String>,
    #[serde(rename = "weatherCode")]
    weather_code: Option<String>,
}

#[derive(Deserialize, Debug)]
struct WttrDay {
    date: Option<String>,
    hourly: Option<Vec<WttrHourlyBlock>>,
}

#[derive(Deserialize, Debug)]
struct WttrResponse {
    weather: Option<Vec<WttrDay>>,
}

fn wwo_code_to_wmo(wwo_code: i32) -> i32 {
    match wwo_code {
        113 => 0,
        116 => 2,
        119 => 3,
        122 => 3,
        143 => 45,
        176 => 61,
        179 => 71,
        182 => 71,
        185 => 56,
        200 => 95,
        227 => 77,
        230 => 75,
        248 => 45,
        260 => 48,
        263 | 266 => 51,
        281 => 56,
        284 => 57,
        293 | 296 => 61,
        299 | 302 => 63,
        305 | 308 => 65,
        311 => 66,
        314 => 67,
        317 => 71,
        320 => 75,
        323 | 326 => 71,
        329 | 332 => 73,
        335 | 338 => 75,
        350 => 77,
        353 => 80,
        356 => 81,
        359 => 82,
        362 => 85,
        365 => 86,
        368 => 85,
        371 => 86,
        374 | 377 => 77,
        386 | 389 => 95,
        392 => 96,
        395 => 99,
        _ => 0,
    }
}

impl WttrProvider {
    fn normalize(&self, raw_data: WttrResponse, start_date: &str, end_date: &str) -> NormalizedWeatherData {
        let mut points = Vec::new();

        if let Some(days) = raw_data.weather {
            for day in days {
                let day_date = match day.date {
                    Some(d) => d,
                    None => continue,
                };
                if day_date.as_str() < start_date || day_date.as_str() > end_date {
                    continue;
                }

                let hourly_blocks = match day.hourly {
                    Some(h) => h,
                    None => continue,
                };
                if hourly_blocks.is_empty() {
                    continue;
                }

                for h in 0..24 {
                    let block_idx = std::cmp::min(h / 3, hourly_blocks.len() - 1);
                    let block = &hourly_blocks[block_idx];

                    let temp_str = block.temp_c.as_deref().unwrap_or("0");
                    let temp = temp_str.parse::<f64>().unwrap_or(0.0);

                    let feels_str = block.feels_like_c.as_deref().unwrap_or(temp_str);
                    let feels_like = feels_str.parse::<f64>().unwrap_or(temp);

                    let prob_str = block.chanceofrain.as_deref().unwrap_or("0");
                    let precip_prob = prob_str.parse::<f64>().unwrap_or(0.0);

                    let precip_str = block.precip_mm.as_deref().unwrap_or("0.0");
                    let precip = precip_str.parse::<f64>().unwrap_or(0.0);

                    let hum_str = block.humidity.as_deref().unwrap_or("0");
                    let humidity = hum_str.parse::<f64>().unwrap_or(0.0);

                    let wind_sp_str = block.wind_speed_kmh.as_deref().unwrap_or("0");
                    let wind_speed = wind_sp_str.parse::<f64>().unwrap_or(0.0);

                    let wind_dir_str = block.wind_dir_degree.as_deref().unwrap_or("0");
                    let wind_direction = wind_dir_str.parse::<f64>().unwrap_or(0.0);

                    let cloud_str = block.cloudcover.as_deref().unwrap_or("0");
                    let cloud_cover = cloud_str.parse::<f64>().unwrap_or(0.0);

                    let code_str = block.weather_code.as_deref().unwrap_or("113");
                    let wwo_code = code_str.parse::<i32>().unwrap_or(113);
                    let wmo_code = wwo_code_to_wmo(wwo_code);

                    points.push(HourlyPoint {
                        time: format!("{}T{:02}:00", day_date, h),
                        temperature: temp,
                        apparent_temperature: feels_like,
                        precipitation_probability: precip_prob,
                        precipitation: precip,
                        humidity,
                        wind_speed,
                        wind_direction,
                        cloud_cover,
                        weather_code: wmo_code,
                        aqi: None,
                        uv_index: None,
                        is_day: None,
                        visibility: None,
                        soil_temperature: None,
                        soil_moisture: None,
                    });
                }
            }
        }

        NormalizedWeatherData {
            provider_name: "wttr.in".to_string(),
            hourly: points,
        }
    }
}

impl BaseWeatherProvider for WttrProvider {
    fn name(&self) -> &str {
        "wttr.in"
    }

    async fn fetch(
        &self,
        ctx: &FetchContext<'_>,
    ) -> Result<NormalizedWeatherData, Box<dyn std::error::Error + Send + Sync>> {
        let url = format!("https://wttr.in/{},{}?format=j1", ctx.lat, ctx.lon);
        let resp = ctx.client.get(&url).send().await?;

        if !resp.status().is_success() {
            return Err(format!("Failed to reach wttr.in: HTTP {}", resp.status()).into());
        }

        let raw_data: WttrResponse = resp.json().await?;
        Ok(self.normalize(raw_data, ctx.start_date, ctx.end_date))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_wttr_interpolation_3h_to_1h() {
        let raw = WttrResponse {
            weather: Some(vec![WttrDay {
                date: Some("2026-05-19".to_string()),
                hourly: Some(vec![
                    WttrHourlyBlock {
                        temp_c: Some("15".to_string()),
                        feels_like_c: Some("14".to_string()),
                        chanceofrain: Some("10".to_string()),
                        precip_mm: Some("0.0".to_string()),
                        humidity: Some("70".to_string()),
                        wind_speed_kmh: Some("12".to_string()),
                        wind_dir_degree: Some("45".to_string()),
                        cloudcover: Some("20".to_string()),
                        weather_code: Some("113".to_string()),
                    },
                    WttrHourlyBlock {
                        temp_c: Some("14".to_string()),
                        feels_like_c: Some("13".to_string()),
                        chanceofrain: Some("20".to_string()),
                        precip_mm: Some("0.2".to_string()),
                        humidity: Some("75".to_string()),
                        wind_speed_kmh: Some("11".to_string()),
                        wind_dir_degree: Some("50".to_string()),
                        cloudcover: Some("40".to_string()),
                        weather_code: Some("116".to_string()),
                    },
                ]),
            }]),
        };

        let provider = WttrProvider;
        let normalized = provider.normalize(raw, "2026-05-19", "2026-05-19");
        assert_eq!(normalized.hourly.len(), 24);

        let pt0 = &normalized.hourly[0];
        assert_eq!(pt0.time, "2026-05-19T00:00");
        assert_eq!(pt0.temperature, 15.0);
    }
}
